packages/fetchai/connections/p2p_noise/noise_node.go

```diff
@@ -0,0 +1,193 @@
+	go func() {
+		for {
+			discover(overlay)
+			time.Sleep(2500*time.Millisecond)
+		}
+	}()
+
+	// Once overlay setup, connect to agent
+	check(agent.Connect())
+
+	// Receive envelopes from agent
+  	go func() {
+    	for envel :=  range(agent.Queue()) {
+			go send(*envel, node, overlay)
+    	}
+  	}()
+
+	// Wait until Ctrl+C or a termination call is done.
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt)
+	<-c
+
+	// Empty println.
+	println()
+}
+
+// Deliver an envelope 
+func  send(envel aea.Envelope, node *noise.Node, overlay *kademlia.Protocol) error {
+    //fmt.Printf("[noise-p2p][debug] Looking for %s...\n", envel.To)
+    ids := overlay.Table().Peers()
+    var dest *noise.ID = nil
+    for _, id := range(ids) {
+      if id.ID.String() == envel.To {
+        dest = &id
+        break
+      }
+    }
+    
+    if dest == nil {
+	    fmt.Printf("[noise-p2p][error] Couldn't locate peer with id %s\n", envel.To)
+      	return errors.New("Couldn't locate peer")
+    }
+
+	fmt.Printf("[noise-p2p][debug] Sending to %s:%s...\n", dest.Address, envel)
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	err := node.SendMessage(ctx, dest.Address, envel)
+	cancel()
+
+	if err != nil {
+		fmt.Printf("[noise-p2p][error] Failed to send message to %s. Skipping... [error: %s]\n",
+			envel.To,
+			err,
+		)
+		return errors.New("Failed to send message")
+	}
+
+	return nil
+}
+
+// Handle envelope from other peers
+func handle(ctx noise.HandlerContext, agent aea.AeaApi) error {
+	if ctx.IsRequest() {
+		return nil
+	}
+
+	obj, err := ctx.DecodeMessage()
+	if err != nil {
+		return nil
+	}
+
+	envel, ok := obj.(aea.Envelope)
+	if !ok {
+		return nil
+	}
+
+	// Deliver envelope to agent
+	fmt.Printf("[noise-p2p][debug] Received envelope %s(%s) - %s\n", ctx.ID().Address, ctx.ID().ID.String(), envel)
+  	agent.Put(&envel)
+
+	return nil
+}
+
+// bootstrap pings and dials an array of network addresses which we may interact with and  discover peers from.
+func bootstrap(node *noise.Node, addresses ...string) error {
+	for _, addr := range addresses {
+		ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+		_, err := node.Ping(ctx, addr)
+		cancel()
+
+		if err != nil {
+			fmt.Printf("[noise-p2p][error] Failed to ping bootstrap node (%s). Skipping... [error: %s]\n", addr, err)
+			return err
+		}
+	}
+	return nil
+}
+
+// discover uses Kademlia to discover new peers from nodes we already are aware of.
+func discover(overlay *kademlia.Protocol) {
+	ids := overlay.Discover()
+
+	var str []string
+	for _, id := range ids {
+		str = append(str, fmt.Sprintf("%s(%s)", id.Address, id.ID.String()))
+	}
+
+	// TOFIX(LR) keeps printing already known peers
+	if len(ids) > 0 {
+		//fmt.Printf("[noise-p2p][debug] Discovered %d peer(s): [%v]\n", len(ids), strings.Join(str, ", "))
+	} else {
+		//fmt.Printf("[noise-p2p][debug] Did not discover any peers.\n")
```

Comment:
is it correct that they are commented?

Reply:
yes, otherwise `discover` will pollute the node's logs.
In subsequent pr, I'll make `discover` ignore already discovered pr.
This is already marked as a `TOFIX`